_posts/2016-04-25-watson-and-rails.md

```diff
@@ -0,0 +1,176 @@
+
+For a Ruby on Rails project, I wanted a user of my web application to be able to upload a PDF and have its contents displayed on the page. Watson's [document conversion](http://www.ibm.com/smarterplanet/us/en/ibmwatson/developercloud/doc/document-conversion/) API looked like the perfect solution. I could input a Word, PDF, or HTML document and convert it to plain text, HTML, or JSON format, which I could then display on the page. Eureka!
+
+Unfortunately, my joy was short-lived. I was new to programming, so imagine my dismay when I discovered that this and other Watson APIs don't have any Ruby documentation. Not only that, but Ruby doesn't natively support uploading a file through an HTTP request.
+
+#### Here's how I made it work:
+
+My app was all about teachers being able to upload assignments so that students could view them and make submissions. Basically, it was for homework.
+
+In my database I have a model called `Assignment` with numerous properties, one of which is the property `file`, which is a string.
+
+When creating a new assignment on the `/assignments/new` page, the file is saved via the following form field:
+
+```ruby
+<div class="file">
+  <%= f.label :file, "File (PDF Only)" %>
+  <%= f.file_field :file %>
+</div>
+```
+The Assignments controller had the necessary `new` and `create` methods to save new assignments to the database, which I won't be covering here.
+
+My converted document would be displayed on the Assignment "show" page (`/assignments/:id`). This is where the call to the Watson API would take place.
+
+In the Assignments controller, I had the following code:
+
+```ruby
+app/controllers/assignments_controller.rb
+
+def show
+  @assignment = Assignment.find(params[:id])
+  # This makes the API request with the file as an argument
+  @document = WatsonApi.new(@assignment.file)
+end
+```
+
+The `WatsonApi` wrapper class would take care of sending the request to the API.
+
+I used the [faraday](https://github.com/lostisland/faraday) gem to format the HTTP request. The `WatsonApi` model contained the following:
+
+```ruby
+app/models/watson_api.rb
+
+require 'json'
+require 'open-uri' #For opening the file when not stored locally
+require 'faraday'
+
+class WatsonApi
+  attr_reader :data
+
+  # Methods called on the Assignment show page when you call WatsonApi.new
+
+  def initialize(doc) #doc argument is the @assignment.file
+    if Rails.env.production?
+      # In production, the file is a url from an S3 bucket
+      @doc = doc.url
+    else
+      # In development, it's a local storage path
+      @doc = doc.path
+    end
+    # This calls the get_doc method whenever we make
+    # a WatsonApi.new request from the show page and sets
+    # the response data equal to the @data variable
+    @data = get_doc
+  end
+
+  def questions
+    # This is for iterating through the returned JSON
+    # to pull out the desired data
+    @data.values[3].map { |item| [item["title"], item["content"][0]["text"]] }
```

Comment:
Suggestions for the comment: "This iterates through the returned JSON and pulls out the desired data, in this case the questions from the Assignment."